File: sudoku_generator.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <chrono>

class SudokuGenerator {
private:
    std::vector<std::vector<int>> grid;
    std::vector<int> numbers;
    std::mt19937 rng;

public:
    SudokuGenerator() : grid(9, std::vector<int>(9, 0)), numbers{1, 2, 3, 4, 5, 6, 7, 8, 9} {
        // Seed the random number generator with current time
        rng.seed(std::chrono::steady_clock::now().time_since_epoch().count());
    }

    bool isValidPlacement(int row, int col, int num) {
        // Check row
        for (int j = 0; j < 9; j++) {
            if (grid[row][j] == num) {
                return false;
            }
        }

        // Check column
        for (int i = 0; i < 9; i++) {
            if (grid[i][col] == num) {
                return false;
            }
        }

        // Check 3x3 box
        int startRow = (row / 3) * 3;
        int startCol = (col / 3) * 3;
        for (int i = startRow; i < startRow + 3; i++) {
            for (int j = startCol; j < startCol + 3; j++) {
                if (grid[i][j] == num) {
                    return false;
                }
            }
        }

        return true;
    }

    bool isValidSudoku() {
        // Check each row
        for (int i = 0; i < 9; i++) {
            std::vector<bool> seen(10, false);
            for (int j = 0; j < 9; j++) {
                if (grid[i][j] == 0 || seen[grid[i][j]]) {
                    return false;
                }
                seen[grid[i][j]] = true;
            }
        }

        // Check each column
        for (int j = 0; j < 9; j++) {
            std::vector<bool> seen(10, false);
            for (int i = 0; i < 9; i++) {
                if (grid[i][j] == 0 || seen[grid[i][j]]) {
                    return false;
                }
                seen[grid[i][j]] = true;
            }
        }

        // Check each 3x3 box
        for (int boxRow = 0; boxRow < 3; boxRow++) {
            for (int boxCol = 0; boxCol < 3; boxCol++) {
                std::vector<bool> seen(10, false);
                for (int i = 0; i < 3; i++) {
                    for (int j = 0; j < 3; j++) {
                        int num = grid[boxRow * 3 + i][boxCol * 3 + j];
                        if (num == 0 || seen[num]) {
                            return false;
                        }
                        seen[num] = true;
                    }
                }
            }
        }

        return true;
    }

    std::vector<int> countNumbers() {
        std::vector<int> counts(10, 0); // indices 0-9, using 1-9
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (grid[i][j] != 0) {
                    counts[grid[i][j]]++;
                }
            }
        }
        return counts;
    }

    bool solve() {
        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                if (grid[row][col] == 0) {
                    // Shuffle numbers for randomness
                    std::shuffle(numbers.begin(), numbers.end(), rng);
                    
                    for (int num : numbers) {
                        if (isValidPlacement(row, col, num)) {
                            grid[row][col] = num;
                            if (solve()) {
                                return true;
                            }
                            grid[row][col] = 0;
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    bool generateSudoku() {
        // Clear the grid first
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                grid[i][j] = 0;
            }
        }

        // Generate the Sudoku
        return solve();
    }

    void printGrid() {
        std::cout << "Generated Sudoku:\n";
        std::cout << "+" << std::string(25, '-') << "+\n";
        
        for (int i = 0; i < 9; i++) {
            if (i % 3 == 0 && i != 0) {
                std::cout << "+" << std::string(25, '-') << "+\n";
            }
            std::cout << "| ";
            for (int j = 0; j < 9; j++) {
                if (j % 3 == 0 && j != 0) {
                    std::cout << "| ";
                }
                std::cout << grid[i][j] << " ";
            }
            std::cout << "|\n";
        }
        std::cout << "+" << std::string(25, '-') << "+\n";
    }

    bool verifySudoku() {
        std::cout << "\nVerification:\n";

        // Check if it's a valid Sudoku
        bool isValid = isValidSudoku();
        std::cout << "✓ Valid Sudoku structure: " << (isValid ? "True" : "False") << "\n";

        // Check number counts
        std::vector<int> counts = countNumbers();
        bool allNine = true;
        for (int i = 1; i <= 9; i++) {
            if (counts[i] != 9) {
                allNine = false;
                break;
            }
        }
        std::cout << "✓ Each number appears exactly 9 times: " << (allNine ? "True" : "False") << "\n";

        // Check no zeros
        bool noZeros = true;
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (grid[i][j] == 0) {
                    noZeros = false;
                    break;
                }
            }
            if (!noZeros) break;
        }
        std::cout << "✓ No empty cells (zeros): " << (noZeros ? "True" : "False") << "\n";

        // Check total cells
        int totalCells = 81;
        std::cout << "✓ Total cells: " << totalCells << " (should be 81)\n";

        return isValid && allNine && noZeros && totalCells == 81;
    }
};

int main() {
    std::cout << "Sudoku Generator\n";
    std::cout << std::string(50, '=') << "\n";

    SudokuGenerator generator;

    // Generate a new Sudoku
    std::cout << "Generating a random valid Sudoku...\n";
    bool success = generator.generateSudoku();

    if (success) {
        generator.printGrid();
        bool verified = generator.verifySudoku();

        if (verified) {
            std::cout << "\n✓ Successfully generated a valid Sudoku!\n";
        } else {
            std::cout << "\n✗ Generated Sudoku failed verification!\n";
        }
    } else {
        std::cout << "✗ Failed to generate a valid Sudoku\n";
    }

    return 0;
}

File: sudoku_generator.py
import random
import copy

class SudokuGenerator:
    def __init__(self):
        self.grid = [[0 for _ in range(9)] for _ in range(9)]
        self.numbers = list(range(1, 10))
    
    def is_valid_placement(self, row, col, num):
        """Check if placing num at (row, col) is valid"""
        # Check row
        if num in self.grid[row]:
            return False
        
        # Check column
        for i in range(9):
            if self.grid[i][col] == num:
                return False
        
        # Check 3x3 box
        start_row = (row // 3) * 3
        start_col = (col // 3) * 3
        for i in range(start_row, start_row + 3):
            for j in range(start_col, start_col + 3):
                if self.grid[i][j] == num:
                    return False
        
        return True
    
    def is_valid_sudoku(self):
        """Check if the entire grid is a valid Sudoku"""
        # Check each row
        for row in self.grid:
            if len(set(row)) != 9 or 0 in row:
                return False
        
        # Check each column
        for col in range(9):
            column = [self.grid[row][col] for row in range(9)]
            if len(set(column)) != 9 or 0 in column:
                return False
        
        # Check each 3x3 box
        for box_row in range(3):
            for box_col in range(3):
                box = []
                for i in range(3):
                    for j in range(3):
                        box.append(self.grid[box_row * 3 + i][box_col * 3 + j])
                if len(set(box)) != 9 or 0 in box:
                    return False
        
        return True
    
    def count_numbers(self):
        """Count occurrences of each number (1-9) in the grid"""
        counts = {i: 0 for i in range(1, 10)}
        for row in self.grid:
            for num in row:
                if num != 0:
                    counts[num] += 1
        return counts
    
    def generate_sudoku(self):
        """Generate a valid filled Sudoku using backtracking with randomization"""
        # Shuffle numbers for randomness
        random.shuffle(self.numbers)
        
        def solve():
            for row in range(9):
                for col in range(9):
                    if self.grid[row][col] == 0:
                        # Try each number in random order
                        random.shuffle(self.numbers)
                        for num in self.numbers:
                            if self.is_valid_placement(row, col, num):
                                self.grid[row][col] = num
                                if solve():
                                    return True
                                self.grid[row][col] = 0
                        return False
            return True
        
        # Clear the grid first
        self.grid = [[0 for _ in range(9)] for _ in range(9)]
        
        # Generate the Sudoku
        if solve():
            return True
        else:
            return False
    
    def print_grid(self):
        """Print the Sudoku grid in a nice format"""
        print("Generated Sudoku:")
        print("+" + "-" * 25 + "+")
        for i, row in enumerate(self.grid):
            if i % 3 == 0 and i != 0:
                print("+" + "-" * 25 + "+")
            row_str = "| "
            for j, num in enumerate(row):
                if j % 3 == 0 and j != 0:
                    row_str += "| "
                row_str += str(num) + " "
            row_str += "|"
            print(row_str)
        print("+" + "-" * 25 + "+")
    
    def verify_sudoku(self):
        """Verify that the generated Sudoku meets all requirements"""
        print("\nVerification:")
        
        # Check if it's a valid Sudoku
        is_valid = self.is_valid_sudoku()
        print(f"✓ Valid Sudoku structure: {is_valid}")
        
        # Check number counts
        counts = self.count_numbers()
        all_nine = all(count == 9 for count in counts.values())
        print(f"✓ Each number appears exactly 9 times: {all_nine}")
        
        # Check no zeros
        no_zeros = all(all(cell != 0 for cell in row) for row in self.grid)
        print(f"✓ No empty cells (zeros): {no_zeros}")
        
        # Check total cells
        total_cells = sum(len(row) for row in self.grid)
        print(f"✓ Total cells: {total_cells} (should be 81)")
        
        return is_valid and all_nine and no_zeros and total_cells == 81

def main():
    """Main function to generate and display a Sudoku"""
    print("Sudoku Generator")
    print("=" * 50)
    
    generator = SudokuGenerator()
    
    # Generate a new Sudoku
    print("Generating a random valid Sudoku...")
    success = generator.generate_sudoku()
    
    if success:
        generator.print_grid()
        generator.verify_sudoku()
        
        print(f"\n✓ Successfully generated a valid Sudoku!")
    else:
        print("✗ Failed to generate a valid Sudoku")

if __name__ == "__main__":
    main()